pkg/noderesource/noderesource.go

```diff
@@ -0,0 +1,340 @@
+	for _, c := range nr.collectorList {
+		if c.Name() == collectorName {
+			continue
+		}
+		tmpState := c.GetLastState()
+		for resourceName := range tmpState {
+			if _, ok := state[resourceName]; ok {
+				state[resourceName] = append(state[resourceName], tmpState[resourceName]...)
+			} else {
+				state[resourceName] = tmpState[resourceName]
+			}
+		}
+	}
+	return state
+}
+
+func (nr *NodeResource) Analyze(state map[string][]MetricTimeSeries) {
+	klog.V(4).Infof("state of NodeResource: %v", state)
+	node, err := nr.nodeLister.Get(nr.nodeName)
+	if err != nil {
+		klog.Errorf("Failed to get node: %v", err)
+		return
+	}
+	nodeCopy := node.DeepCopy()
+	effectDataSourceNameMap := make(map[string]string)
+	lowerLimitResourceNameMap := make(map[string]float64)
+	//获取指标上限值
+	for resourceName, mTsList := range state {
+		for _, metricTimeSeries := range mTsList {
+			if metricTimeSeries.DataSourceName != GetRealtimeCollectorName() {
+				continue
+			}
+			for _, timeSeries := range metricTimeSeries.TimeSeriesList {
+				for _, sample := range timeSeries.Samples {
+					if lowerLimitResourceNameMap[resourceName] < sample.Value*0.9 {
+						lowerLimitResourceNameMap[resourceName] = sample.Value * 0.9
+					}
+				}
+			}
+		}
+	}
+	for resourceName, mTsList := range state {
+		var miniIdle, nextIdle *float64
+		for _, metricTimeSeries := range mTsList {
+			for _, timeSeries := range metricTimeSeries.TimeSeriesList {
+				for _, sample := range timeSeries.Samples {
+					if sample.Timestamp == 0 {
+						continue
+					}
+					nextIdle = &sample.Value
+					if lowerLimitResource, ok := lowerLimitResourceNameMap[resourceName]; ok {
+						if *nextIdle < lowerLimitResource {
+							klog.V(4).Infof("The predicted sample of resource %s is lower than the lowerLimitValue, sampleValue: %f, lowerLimitValue: %f", resourceName, nextIdle, lowerLimitResource)
+							continue
+						}
+					}
+					if miniIdle == nil {
+						miniIdle = nextIdle
+						effectDataSourceNameMap[resourceName] = metricTimeSeries.DataSourceName
+					}
+					if *miniIdle > *nextIdle {
+						miniIdle = nextIdle
+						effectDataSourceNameMap[resourceName] = metricTimeSeries.DataSourceName
+					}
+				}
+			}
+		}
+		nr.BuildNodeStatus(resourceName, *miniIdle, nodeCopy)
+	}
+
+	if !equality.Semantic.DeepEqual(&node.Status, &nodeCopy.Status) {
+		var jsonPatch []byte
+		var oldJson []byte
+		var newJson []byte
+		var err error
+		effectDataSources, _ := json.Marshal(effectDataSourceNameMap)
+		nr.recorder.Event(utils.GetNodeRef(nr.nodeName), v1.EventTypeNormal, "RecommendExtendResource", fmt.Sprintf("Recommend Node Extend Resource Success: %s", effectDataSources))
+		oldJson, err = json.Marshal(node)
+		if err == nil {
+			newJson, err = json.Marshal(nodeCopy)
+			if err == nil {
+				jsonPatch, err = strategicpatch.CreateTwoWayMergePatch(oldJson, newJson, nodeCopy)
+				klog.V(4).Infof("jsonPatch: %s", jsonPatch)
+			}
+		}
+
+		if err != nil {
```

Comment:
The logic is to use jsonPatch with PatchStatus first, if err,  use UpdateStatus?
Why not use UpdateStatus directly?